api/lib/infrastructure/utils/redis-rate-limiter.js

```diff
@@ -0,0 +1,21 @@
+const redis = require('redis');
+const { promisify } = require('util');
+const logger = require('../logger');
+const settings = require('../../config');
+
+class RedisRateLimiter {
+  constructor() {
+    if (settings.rateLimit.redisUrl) {
+      this._client = redis.createClient(settings.rateLimit.redisUrl);
+      this._client.scriptAsync = promisify(this._client.script).bind(this._client);
+      this._client.evalshaAsync = promisify(this._client.evalsha).bind(this._client);
+      this._client.on('error', (err) => logger.warn({ redisClient: 'rate-limit', err }, 'Error encountered'));
+      return this._client;
+    } else {
+      this._client = null;
+    }
+    return this._client;
+  }
+}
+
+module.exports = new RedisRateLimiter();
```

Comment:
Pourquoi en faire une classe, qui n'a qu'un constructeur et qui retourne autre chose que le traditionnel `this` ? 🤔

Reply:
Que vois-tu comme autre solution ? Je suis preneur !

Reply:
Je vois pas le problème avec la fonction précédente : `createRedisRateLimit`

Reply:
Je l'ai indiqué dans la description de la PR, mais ce n'est probablement pas assez clair

On a besoin de singleton pour accéder au client, et pouvoir invoquer la méthode `quit()`. 

Dans le cas que tu abordes, le rate-limiter, le module expose un constructeur `createRedisRateLimit` au lieu d'une instance. En conséquence, l'appel de ce constructeur `createRedisRateLimit(config.rateLimit.redisUrl)` dans `plugins.js` crée une instance, mais cette instance n'est pas accessible en dehors de `plugin.js`, notamment dans `api/index.js` où l'on veut déconnecter le client. On ne peut donc pas déconnecter le client.

Reply:
Corrigé dans https://github.com/1024pix/pix/pull/4666#discussion_r958128876